src/tools/miri/src/concurrency/data_race.rs

```diff
@@ -732,6 +732,7 @@ impl VClockAlloc {
                 | MiriMemoryKind::Tls,
             )
             | MemoryKind::CallerLocation => (VTimestamp::ZERO, VectorIdx::MAX_INDEX),
+            MemoryKind::StaticRoot(_) => unreachable!("only used in const eval, but can't be a machine memory kind"),
```

Comment:
The `Machine` trait has a way  for machine instances to add more memory kinds. This looks like the const-eval machine should use that to add `StaticRoot`, rather than add it for everyone?

Reply:
We already have before_memory_read... you could just know the AllocId that is "banned" and reject all reads from that? Then you don't even need to use a new memory kind.

Hm, but for zero-sized reads we might not even generate a `before_memory_read`... how does your current scheme catch those? Do we have a test for the zero-sized recursive static case? Looking at `check_and_deref_ptr` I don't see where we'd ever call `get_alloc_raw`...

Reply:
Ah, I found it, it's in `get_ptr_alloc`.

So, alternative proposal: new hook in `get_alloc_raw` that is passed the `AllocId`, and the CtfeMachine simply has a `write_only_allocation: Option<AllocId>` to ban reads from that allocation. The main issue is finding a name for that hook.^^ Maybe we want a `before_access_machine_alloc` hook that mirrors `before_access_global`?